Failure detector improvements

MaEtUgR opened the pull request:
**Describe problem solved by this pull request**
I found:
1. Commander contained two very similar conditions to process failure detector output
1.  Failure detector used different naming conventions
1. Failure detector contained extensive logic to keep track of the state getting updated or not. This can be done by just comparing the one-byte state before and after.
1.  Failure detector reactions to attitude tracking checks are not independent of reactions to ESC checks

**Test data / coverage**
No tests done.

MaEtUgR commented:
@jkflying I found two more small improvements, last two commits.

jkflying commented on src/modules/commander/failure_detector/FailureDetector.cpp:
```diff
@@ -62,12 +62,8 @@ bool FailureDetector::update(const vehicle_status_s &vehicle_status)
 		_status &= ~(FAILURE_ROLL | FAILURE_PITCH | FAILURE_ALT | FAILURE_EXT);
 	}
 
-	if (_esc_status_sub.updated()) {
```
So the difference here is that we run it on every iteration, even if there is no status update. Will that cause an issue, maybe if the first update hasn't arrived yet?

MaEtUgR replied:
FYI This question is still valid since even though this line got changed it was only the naming.
Answer: I did not remove the condition on the esc_status being updated but I moved it into the `updateEscsStatus()` function such that consistently `updateXXX()` gets called for all the checks and the check itself decides what to depend on e.g. topic update. See https://github.com/PX4/Firmware/pull/15457/files#diff-84f12e22ed352e6c0b6b55e22b6f0628R160

MaEtUgR commented:
I rebased, added the last commit that improves the latest findings of messaging. I'm open to review. The commits should explain a lot better what has been done compared to the combined diff.

dagar commented on src/modules/commander/failure_detector/FailureDetector.cpp:
```diff
@@ -47,50 +47,42 @@ FailureDetector::FailureDetector(ModuleParams *parent) :
 
 		if (_param_fd_ext_ats_en.get()) {
-			updated |= updateExternalAtsStatus();
+			updateExternalAtsStatus();
```
Why should "external ATS" be dependent on attitude stabilized?
@dakejahl

dakejahl replied:
When flying manually in attitude or rates mode you probably don't want your ATS device to trigger the parachute, since you are in manual control anyway and you are probably doing some sketchy maneuvers (as pilots like to do).

dakejahl replied:
I could see it go either way though

dagar replied:
Ok, I thought this was simply telling you the external thing triggered a parachute and you should now shut off your motors so it actually works?

Either way maybe we should add a comment.

dakejahl replied:
Oh. Yeah you're right, that's totally how it works. Yes it should be changed.

MaEtUgR replied:
I didn't change that logic. Should I? Can we do a follow-up pr?

dagar replied:
Whatever's easier.

MaEtUgR replied:
Let's do a follow up to not derail.

dagar commented on src/modules/commander/Arming/PreFlightCheck/checks/failureDetectorCheck.cpp:
```diff
@@ -45,17 +45,30 @@ bool PreFlightCheck::failureDetectorCheck(orb_advert_t *mavlink_log_pub, const v
+
 			if (status.failure_detector_status & vehicle_status_s::FAILURE_ROLL) {
-				mavlink_log_critical(mavlink_log_pub, "Preflight Fail: Roll failure detected");
+				message = "Roll";
```
This is fine, but in general I think I'd prefer to not be manipulating strings at all. We need to move beyond that entirely for a number of reasons (events interface, localization, flash usage, etc).

dagar replied:
More specifically one reason it's preferable to leave each string together as a single literal is for the possibility to tokenize later and replace the actual string literal with a more compact binary token.

MaEtUgR replied:
I see, I actually did this to safe flash since the strings contain quite some redundancy. I'm happy to follow your suggestions. I can change it back if you like.

MaEtUgR replied:
I changed it in "failureDetectorCheck: remove string manipulation". It uses 168Bytes more flash :man_shrugging:

dagar commented:
Generally looks good and overlaps quite a bit with the changes I made in https://github.com/PX4/Firmware/pull/15650.

bresch commented on src/modules/commander/Commander.cpp:
```diff
@@ -2226,56 +2226,42 @@ Commander::run()
-					if (status.failure_detector_status & vehicle_status_s::FAILURE_ARM_ESC) {
+					if (hrt_elapsed_time(&time_at_arm) < 500_ms) {
 						arm_disarm(false, true, &mavlink_log_pub, arm_disarm_reason_t::FAILURE_DETECTOR);
 						_status_changed = true;
```
`_status_changed` is already set to `true` above

MaEtUgR replied:
Let me check

MaEtUgR replied:
I corrected it in "Commander: remove duplicate flag changes". Thanks for the catch :+1:

bresch commented on src/modules/commander/Commander.cpp:
```diff
@@ -2226,56 +2226,42 @@ Commander::run()
+						// This handles the case where something fails during the early takeoff phase
 						armed.lockdown = true;
 						_lockdown_triggered = true;
 						_status_changed = true;
```
same

bresch commented on src/modules/commander/Commander.cpp:
```diff
@@ -2226,56 +2226,42 @@ Commander::run()
-					set_tune_override(TONE_PARACHUTE_RELEASE_TUNE);
+						armed.force_failsafe = true;
+						_flight_termination_triggered = true;
+						_status_changed = true;
```
same

bresch reviewed (changes requested):
Just a small mistake. All good for the rest.

MaEtUgR commented:
> Generally looks good and overlaps quite a bit with the changes I made in #15650.

That's why I brought it up in the dev call. I hope it's not too bad to rebase.

dagar commented:
> > Generally looks good and overlaps quite a bit with the changes I made in #15650.
> 
> That's why I brought it up in the dev call. I hope it's not too bad to rebase.

No problem, it should be good. To move https://github.com/PX4/Firmware/pull/15650 forward I need to get all those unrelated changes out of it.

MaEtUgR commented:
I rebased on master and addressed the comments if you could check.
